fs/zipfs.py

```diff
@@ -207,63 +293,63 @@ def _directory(self):
-                    }
-                }
-            else:
-                modified_epoch = datetime_to_epoch(
-                    datetime(*zip_info.date_time)
-                )
-                raw_zip_info = {
-                    k: getattr(zip_info, k)
-                    for k in dir(zip_info)
-                    if (not k.startswith('_') and
-                        not callable(getattr(zip_info, k)))
-                }
-                raw_info = {
-                    "basic":
-                    {
-                        "name": basic_info.name,
-                        "is_dir": basic_info.is_dir,
-                    },
-                    "details":
-                    {
-                        "size": zip_info.file_size,
-                        "type": int(
-                            ResourceType.directory
-                            if basic_info.is_dir else
-                            ResourceType.file
-                        ),
-                        "modified": modified_epoch
-                    },
-                    "zip": raw_zip_info
-                }
+            raw_info["basic"] = {
+                "name": basic_info.name,
+                "is_dir": basic_info.is_dir,
+            }
+
+            if not {"details", "access", "zip"}.isdisjoint(namespaces):
+                zip_name = self._path_to_zip_name(path)
+                try:
+                    zip_info = self._zip.getinfo(zip_name)
+                except KeyError:
+                    # Can occur if there is an implied directory in the zip
+                    pass
+                else:
+                    if "details" in namespaces:
+                        raw_info["details"] = {
+                            "size": zip_info.file_size,
+                            "type": int(
+                                ResourceType.directory
+                                if basic_info.is_dir else
+                                ResourceType.file
+                            ),
+                            "modified": datetime_to_epoch(
+                                datetime(*zip_info.date_time)
+                            )
+                        }
+                    if "zip" in namespaces:
+                        raw_info["zip"] = {
+                            k: getattr(zip_info, k)
+                            for k in dir(zip_info)
+                            if (not k.startswith('_') and
+                                not callable(getattr(zip_info, k)))
+                        }
+                    if "access" in namespaces:
+                        # check the zip was created on UNIX to get permissions
+                        if zip_info.create_system == 3:
+                            mode = zip_info.external_attr >> 16 & 0xFFF
+                            raw_info["access"] = {
+                                "permissions": Permissions(
+                                    mode=mode or 0o600  # default permission
```

Comment:
If the permissions bits are 0 does that indicate default permissions in a zip file? Because I think 0 is a valid (if unusual) value for unix permissions.

Reply:
I checked by doing this:
```python
>>> import zipfile
>>> with zipfile.ZipFile('test.zip', 'w') as z:
...    info = zipfile.ZipInfo('test.txt')
...    z.writestr(info, '')
```
and then 
```
$ zipinfo test.zip
Archive:  test.zip
Zip file size: 114 bytes, number of entries: 1
?rw-------  2.0 unx        0 b- stor 80-Jan-01 00:00 test.txt
1 file, 0 bytes uncompressed, 0 bytes compressed:  0.0%
```

So I think actually having external attr as `0` means that there are no permission defined, which will result in `0o600` being applied when extracting on Unix. So it may be a better idea to only add the `access/permissions` info if external attr is the permissions are defined, and let the user do what he wants on no permissions.

Permissions of zip objects are a mess, since zip file don't store the uid/gid of the file owners, so even with permissions mode they are rendered pretty useless.

Reply:
Fair enough.

I've given thought to implementing some kind of custom PyFilesystem archive format that could capture everything that an FS object can expose. But it sounds like a lot of work, and there are probably bigger fish to fry...

Reply:
This could be interesting indeed !